### PR TITLE
[API] As a user, I can logout

### LINKED ISSUE #11: [API] As a user, I can logout
## Why

Support user logout functionality via API

## Acceptance Criteria
- [ ]  User can able to logout via API
- [ ]  Expose proper response if the logout request is unsuccessful
- [ ]  Unit test

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/google-scraper/projects/1#card-50542063

## What happened 👀

User can logout by logout requesting with valid credential

## Insight 📝

N/A

### Successful logout response
Screen Shot 2021-04-01 at 10 57 24 AM
### Wrong client credential response
Screen Shot 2021-04-01 at 11 12 28 AM
### Empty token response
Screen Shot 2021-04-01 at 11 02 39 AM

### CODE REVIEW

File: controllers/api/v1/token.go [hunk 1]
```diff
@@ -36,6 +37,31 @@ func (c *Token) Create() {
 	c.serveJSON(&tokenResponse)
 }
 
+func (c *Token) Revoke() {
+	err := c.authenticateClient()
```
Comment (junan) [inline controllers/api/v1/token.go:38]: I found in the [Doorkeeper](https://github.com/doorkeeper-gem/doorkeeper/blob/main/app/controllers/doorkeeper/tokens_controller.rb#L5) that it validates client credentials also for revoking token. So I validated it here too.

File: tests/oauth_token_fabricator.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package tests
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-oauth2/oauth2/v4/models"
```
Comment (Lahphim) [inline tests/oauth_token_fabricator.go]: @junan Move this to the group 🙌

File: tests/oauth_token_fabricator.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/go-oauth2/oauth2/v4"
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo"
+)
+
+func FabricateOAuthToken(client *models.Client) oauth2.TokenInfo {
+	info := &models.Token{
+		ClientID:        client.GetID(),
+		Access:          uuid.New().String(),
+		AccessCreateAt:  time.Now(),
```
Comment (Lahphim) [inline tests/oauth_token_fabricator.go]: @junan Should we use `time.Now().Local()` ?
```suggestion
		AccessCreateAt:  time.Now().Local(),
```

File: routers/router.go [hunk 4]
```diff
@@ -3,8 +3,9 @@ package routers
 import (
 	"google-scraper/controllers"
 
-	"github.com/beego/beego/v2/server/web"
 	apiv1 "google-scraper/controllers/api/v1"
```
Comment (Lahphim) [inline routers/router.go:5]: @junan Should we group this?
```golang
"google-scraper/controllers"
apiv1 "google-scraper/controllers/api/v1"

...
```

File: controllers/api/v1/token_test.go [hunk 5]
```diff
@@ -191,6 +193,175 @@ var _ = Describe("TokenController", func() {
...
+				user := FabricateUser("John", email, password)
+
+				form := url.Values{
+					"client_id":     {client.ID},
+					"client_secret": {client.Secret},
+					"token":         {token.GetAccess()},
+				}
+				body := strings.NewReader(form.Encode())
+
+				response := MakeAuthenticatedRequest("POST", "/api/v1/revoke", body, &user)
```
Comment (Lahphim) [inline controllers/api/v1/token_test.go]: @junan Why do we have to make a request with `&user`?

Comment (junan) [reply controllers/api/v1/token_test.go]: Side effect of copy/paste from other test 🙈

File: controllers/api/v1/token.go [hunk 6]
```diff
@@ -34,3 +35,43 @@ func (c *Token) Create() {
 
 	c.serveJSON(&tokenResponse)
 }
+
+func (c *Token) Revoke() {
+	err := c.authenticateClient()
+	if err != nil {
+		c.renderError(err, http.StatusUnauthorized)
+		return
```
Comment (Lahphim) [inline controllers/api/v1/token.go]: @junan I saw one of Hoang's PR use `c.StopRun()` inside `c.renderError()`  function, so we can omit `return` after calling `c.renderError()` function 👍
It looks really clean to have it

**Review by Lahphim — APPROVED** [review]

**Comment by Lahphim** [comment]
@junan Later on can you update a little bit detail about the screenshot?
because it I quite a little bit confuse about purposing of the them

**Review by gutakk — APPROVED** [review]

**Pull request merged**